src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/FileBasedProgramsProjectSystem.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Composition;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.HostWorkspace.ProjectTelemetry;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.ProjectSystem;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.Composition;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
+
+[Export(typeof(FileBasedProgramsProjectSystem)), Shared]
+internal sealed class FileBasedProgramsProjectSystem : LanguageServerProjectLoader
+{
+    private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public FileBasedProgramsProjectSystem(
+        LanguageServerWorkspaceFactory workspaceFactory,
+        IFileChangeWatcher fileChangeWatcher,
+        IGlobalOptionService globalOptionService,
+        ILoggerFactory loggerFactory,
+        IAsynchronousOperationListenerProvider listenerProvider,
+        ProjectLoadTelemetryReporter projectLoadTelemetry,
+        ServerConfigurationFactory serverConfigurationFactory,
+        BinlogNamer binlogNamer)
+            : base(
+                workspaceFactory.FileBasedProgramsProjectFactory,
+                workspaceFactory.TargetFrameworkManager,
+                workspaceFactory.ProjectSystemHostInfo,
+                fileChangeWatcher,
+                globalOptionService,
+                loggerFactory,
+                listenerProvider,
+                projectLoadTelemetry,
+                serverConfigurationFactory,
+                binlogNamer)
+    {
+    }
+
+    public async Task OpenFileBasedProgramAsync(string filePath, string virtualProjectContent, TextDocument? primordialDoc)
+    {
+        using (await _gate.DisposableWaitAsync())
+        {
+            ProjectsToLoadAndReload.AddWork(new VirtualProjectToLoad(filePath, virtualProjectContent, ProjectGuid: null, ReportTelemetry: true));
+
+            // Wait for the in progress batch to complete and send a project initialized notification to the client.
+            await ProjectsToLoadAndReload.WaitUntilCurrentBatchCompletesAsync();
+
+            if (primordialDoc is { })
+            {
+                // Now that the real fbp project is ready, remove the primordial project
+                // TODO(fbp): this flow seems to result in syntax diags from the primordial project showing up, and only disappearing when the doc is edited.
+                ProjectFactory.Workspace.SetCurrentSolution(sln => sln.RemoveProject(primordialDoc.Project.Id), WorkspaceChangeKind.ProjectRemoved);
+            }
```

Comment:
@jasonmalinowski this is what I was trying to describe offline; in order to have this one workspace fulfill both the original "sync backstop" task as well as "minimal set of references, etc." task, there is a need to make a 'primordial' equivalent to the current misc workspace project, async load the virtual project, then remove the primordial project when ready.

Reply:
>  there is a need to make a 'primordial' equivalent to the current misc workspace project, async load the virtual project, then remove the primordial project when ready.

Do we need the primordial stage to even have any references/options/etc. if we're likely to get to the 'good/loaded' stage shortly?  Basically, it seems fine to me if you just get nothing semantic, and only syntax.  After all, we're going to consider the project 'not fully loaded' and so will not have diags or anything like that :)

Reply:
> Do we need the primordial stage to even have any references/options/etc. if we're likely to get to the 'good/loaded' stage shortly?

That's right, what I meant by "equivalent to current misc workspace project" is that the primordial project doesn't have any real references, meaningful compile/parse options etc.

The bug is that at the point where the primordial is unloaded and the real one is loaded in, there are suddenly syntax diags present, which only make sense if the primordial's parse options were used for parsing. And the diags disappear upon making further edits.

It's specifically that the primordial is not using `/features:FileBasedProgram`, and the unexpected syntax errors are on `#:`s which error when that feature flag is not set.

Reply:
> It's specifically that the primordial is not using /features:FileBasedProgram

Ah yeah.  We should jsut fix that.  THe primordial state should be in the maximally allowing state.

That said, we should not be getting semantic diagnostics for files in the primordial state.  So i don't know why we're getting diagnostics for `#:` at all.  Those should just be semantic diagnostics, as structurally we know exactly what it means.  And we certainly don't want to be repeatedly incrementally reparsing files that contain these totally normal directives.

Reply:
@jjonescz we probably want to move ‘#:’ and ‘#!’ diagnostics to binding for the above reasons. I don’t think there is any point those nodes are “bound” but perhaps there is some visit pass which could hit them and report the diagnostics in some convenient way.

Reply:
I’m also currently pursuing Jason’s suggestion to have the FBPProjectSystem own the process of creating the “initial project”, and then  having it do an update on the same project, instead of inserting a new one and deleting the old one as we are currently doing here.

Reply:
FYI, if binding has an entry point that starts with the tree (or the root) this should be trivial. We can already walk the green tree trivial and cheaply thanks to things like HasDirective.  So just walk that down as appropriate.  

This can be at the start of binding or the end.  As most files don't have directives, and most directives are high level (around some high level construct), this walks exceptionally fast. 

The ide has code for this ourselves when walking and determining directives, and it has never shown up as costing anything.

Reply:
Wouldn't it be easier to filter out this specific diagnostic in the IDE? Or pass the feature flag during parsing as was suggested.

It sounds like overkill to add a separate walking pass to binding just for this diagnostic.

And I'm not fully convinced compiler shouldn't report this diagnostic during parsing. In general I think the compiler reports diagnostics where convenient and usually as soon as possible. We just try to avoid *conditional parsing* but that's not what's happening here - we parse those always as IgnoredDirectiveSyntax regardless of the feature flag.

Reply:
The motivation to move the `#:` feature error to binding would just be for incremental parsing. Since nodes with syntax errors are not parsed incrementally, it is thought to be better to only report errors during parsing when something is genuinely syntactically malformed. So, with recent features for example, even if we can tell from syntax that a new feature is being used, and it's not supported in current language version, we wait to report that until binding.

It's not clear to me if only the parsing of the directives themselves is impacted or if the node with such directives in their trivia list are also impacted.

The issue described with the diagnostic appearing in IDE is a symptom of a separate underlying problem I am currently working on.

Reply:
> with recent features for example, even if we can tell from syntax that a new feature is being used, and it's not supported in current language version, we wait to report that until binding.

I see, that seems fine, but for `#:` directives, I don't think we have a place in binding we could report the diagnostic at. So I'm wondering if implementing that would require some new "pass" which would cause a perf hit for all programs. But if IDE could instead just parse with `/features:FileBasedPrograms` that seems like a good solution?

Reply:
> I don't think we have a place in binding we could report the diagnostic at. 

Is there not a top level place where the compilation unit is bound?

Reply:
> which would cause a perf hit for all programs.

There should be practically no hit. You just walk the green tree, only descending into nodes with the HasDirective bit set.

Reply:
> Is there not a top level place where the compilation unit is bound?

Sure, sorry, I meant we don't have a pre-existing binding place per each `#:` directive - so we likely need to create our own pass to scan for the directives.

> There should be practically no hit. You just walk the green tree, only descending into nodes with the HasDirective bit set.

Makes sense, I can do that.

Is there a reason you cannot parse with `/features:FileBasedPrograms` though? That seems much simpler.

Reply:
I don't quite know what that flag is, it how it changes things. Tbh, I'm not sure why there's a flag here for this at all :-)

Reply:
Again, the motivation for changing where to report the diagnostic, isn’t so we can fix a functional issue with the file based programs IDE support. It’s just so that if the #:s are encountered in a scenario where the errors *are* expected, then incremental parsing still works.

Reply:
> I don't quite know what that flag is, it how it changes things. Tbh, I'm not sure why there's a flag here for this at all :-)

When this flag is not present, `#:` directives cause errors to alert the user that they are not supported (they do nothing) in project-based programs. The flag is set automatically during `dotnet run file.cs`.

> Again, the motivation for changing where to report the diagnostic, isn’t so we can fix a functional issue with the file based programs IDE support. It’s just so that if the #:s are encountered in a scenario where the errors _are_ expected, then incremental parsing still works.

Thanks, I missed that. Makes more sense then.

Reply:
No reason, it could be a parse option. I'm also not sure what's the difference. Apart from the fact that feature flags don't undergo API review (but maybe they should?)

Reply:
I think a Features flag is an appropriate control mechanism for the current phase of development. At any rate would like to encourage that discussion to move to its own issue if you would like to continue it.